src/fabric_quality_factor.erl

```diff
@@ -0,0 +1,213 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+
+%% Calculate the shard sync quality factor for a single node/shard copy. This
+%% is the number of updates to other shard copies not present on TargetNode
+%% ShardName  - e.g. <<"shards/c0000000-ffffffff/ksnavely/motohacker.1368999314">>
+%% TargetNode - e.g. 'dbcore@db11.julep.cloudant.net'
+shard_sync_quality_factor(not_found, _TargetNode) ->
+    not_found;
+shard_sync_quality_factor(ShardName, TargetNode) ->
+    DbName = mem3:dbname(ShardName),
+    Nodes = nodes_for_shard(ShardName, DbName),
+
+    %% Find the upper bound of updates to other shard copies which are missing
+    %% from TargetNode
+    MissingUpdates = [shard_missing_updates(Source, TargetNode, ShardName) ||
+        Source <- Nodes, Source =/= TargetNode],
+
+    %% The quality factor is the sum of possible missing updates
+    lists:sum(MissingUpdates).
+
+%% Fetch a single checkpoint doc given target, source, and shard name
+%% ShardName  - e.g. <<"shards/c0000000-ffffffff/ksnavely/motohacker.1368999314">>
+%% TargetNode - e.g. 'dbcore@db14.julep.cloudant.net'
+%% SourceNode - e.g. 'dbcore@db14.julep.cloudant.net'
+local_checkpoint(TargetNode, SourceNode, ShardName) ->
+    {ok, Db} = rpc:call(SourceNode, couch_db, open, [ShardName, [?ADMIN_CTX]]),
+    UUID = couch_db:get_uuid(Db),
+    {_LocalId, Doc} = mem3_rpc:load_checkpoint(TargetNode, ShardName, SourceNode, UUID),
+    Doc.
+
+%% Find upper bound of updates to ShardName on SourceNode that are missing
+%% from TargetNode, which also holds a copy of ShardName
+%% SourceNode - e.g. 'dbcore@db9.julep.cloudant.net'
+%% TargetNode - e.g. 'dbcore@db11.julep.cloudant.net'
+%% ShardName  - e.g. <<"shards/c0000000-ffffffff/ksnavely/motohacker.1368999314">>
+shard_missing_updates(SourceNode, TargetNode, ShardName) ->
+    %% The number of missing updates is the difference in sequence numbers.
+    shard_seq(SourceNode, ShardName) - shard_checkpoint_seq(SourceNode, TargetNode, ShardName).
+
+%% Fetch the last replicated source seq from the replication checkpoint doc.
+%% SourceNode - e.g. 'dbcore@db9.julep.cloudant.net'
+%% TargetNode - e.g. 'dbcore@db14.julep.cloudant.net'
+%% ShardName  - e.g. <<"shards/c0000000-ffffffff/ksnavely/motohacker.1368999314">>
+shard_checkpoint_seq(SourceNode, TargetNode, ShardName) ->
+    CheckpointDoc = local_checkpoint(TargetNode, SourceNode, ShardName),
+    {Body} = element(4, CheckpointDoc),
+    couch_util:get_value(<<"seq">>, Body).
+
+%% Get the seq of a shard replica from a specific node
+%% Node - e.g. 'dbcore@db14.julep.cloudant.net'
+%% ShardName  - e.g. <<"shards/c0000000-ffffffff/ksnavely/motohacker.1368999314">>
+shard_seq(Node, ShardName) ->
+    {ok, Db} = rpc:call(Node, couch_db, open, [ShardName, [?ADMIN_CTX]]),
+    couch_db:get_update_seq(Db).
+
+nodes_for_shard(Shard, DbName) ->
+    lists:map(fun(S1) -> element(3, S1) end,
+        lists:filter(fun(S) ->
+            case element(2, S) of
```

Comment:
Same comment about records and accessors here -- `S1#shard.node` and `S#shard.name` would be preferable here.